Better errors for non literal msgid

vadz opened the pull request:
@lanurmi What do you think of the first commit here?

The second one seems like a minor simplification to me, but we can omit it if you prefer to keep the namespaces, I mostly want to improve the story with the error messages here.

vadz committed:
Improve error message when non-literal message IDs are used

Trigger a static_assert pointing to the comment explaining the problem
and possible fixes/workarounds.

See #24957.

vadz committed:
Don't bother with namespace for wxNO_REQUIRE_LITERAL_MSGIDS

It's not really clear what does this buy us, compared to just plain
conditional compilation.

vadz commented:
Hmm, maybe we should use an URL instead of referencing a comment? It would be directly clickable then and could point to a more detailed explanation...

lanurmi reviewed (commented):
Looks good. A URL could be nice in practice, though it's best to also keep the explanation in comments so it's there to be found offline, too.

vadz commented on include/wx/translation.h:
```diff
@@ -342,13 +342,41 @@ namespace wxTransImplStrict
+    functions. This most likely indicates a bug in your program which is now
+    detected when it was silently ignored before and should be fixed by
+    changing the code to use string literals.
+
```
Added an example with the fix, thanks.

vadz commented on include/wx/translation.h:
```diff
@@ -342,13 +342,41 @@ namespace wxTransImplStrict
+    this is *NOT* recommended, as the problematic strings probably won't be
+    translated, because they won't have been extracted by xgettext in the first
+    place and the right thing to do remains to fix the code instead.
+
```
I don't think we should advise this, as long as they're changing their code, it's better to fix it properly. The `wxNO_REQUIRE_LITERAL_MSGIDS` escape hatch is just for the people who get many such errors and don't want to change anything and just want to compile the existing code.

vadz committed:
Add test check that compiling code using invalid msgids fails

Verify that the expected compilation error is triggered when using _()
translation macro incorrectly.

vadz commented:
See also https://github.com/wxWidgets/website/pull/97

vadz commented:
Amended version merged in the commit above.
